### PR TITLE
Provide axis('equal') for Axes3D (replaces PR #23017)

### PR DESCRIPTION
**Pull request by scottshambaugh**
## PR Summary
This replaces https://github.com/matplotlib/matplotlib/pull/23017. As requested in the thread I rebased and added the What's New entry. I _think_ that @patquem will retain the credit for their work in the first commit, please let me know if I did this wrong.

Closes https://github.com/matplotlib/matplotlib/issues/22570

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [x] New features are documented, with examples if plot related.
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 1]
```diff
@@ -321,14 +315,27 @@ def set_aspect(self, aspect, adjustable=None, anchor=None, share=False):
...
             aspect, adjustable=adjustable, anchor=anchor, share=share)
 
+        if aspect == 'equal':
+            self.set_box_aspect((1., 1., 1.))
+            v_intervals = np.vstack((self.xaxis.get_view_interval(),
+                                     self.yaxis.get_view_interval(),
+                                     self.zaxis.get_view_interval())).T
+            mean = np.mean(v_intervals, axis=0)
+            delta = np.max(np.max(v_intervals, axis=0) -
+                           np.min(v_intervals, axis=0))
```
Comment (tfpf) [inline lib/mpl_toolkits/mplot3d/axes3d.py:333]: Can we achieve the same effect using `np.ptp`?

Also, using `axis=0` on the transpose should be the same as using `axis=1` on the original?

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 2]
```diff
@@ -321,14 +315,27 @@ def set_aspect(self, aspect, adjustable=None, anchor=None, share=False):
...
+
+            self.set_xlim3d(mean[0] - delta / 2., mean[0] + delta / 2.)
+            self.set_ylim3d(mean[1] - delta / 2., mean[1] + delta / 2.)
+            self.set_zlim3d(mean[2] - delta / 2., mean[2] + delta / 2.)
```
Comment (tfpf) [inline lib/mpl_toolkits/mplot3d/axes3d.py:337]: Looks like these might change the limits set by the user?

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 1 above)
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: Yeah, those do clean things up

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 2 above)
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axes3d.py:336]: Yes, this must necessarily happen to change the aspect ratios. If the user sets limits afterwards, that will also break the equal aspect ratios

**Comment by tfpf** [comment]
> If the user sets limits afterwards, that will also break the equal aspect ratios

Oh, I see now: the aspect ratio will be set for the current view. Makes perfect sense!

There appears to be connection between setting the box aspect to `(1, 1, 1)` and the new dimensions being `(delta, delta, delta)`. Hacking it might preserve the limits … If I understand `set_box_aspect` correctly,
```
v_internals = np.array((self.xaxis.get_view_interval(),
                        self.yaxis.get_view_interval(),
                        self.zaxis.get_view_interval()))
deltas = np.ptp(v_intervals, axis=1)
self.set_box_aspect(deltas)
```
should do the trick. (Unable to test this right now, so I'll let somebody else decide if this is preferable.)

**Comment by scottshambaugh** [comment]
@tfpf I actually quite like the result with your idea there:
image

I had another idea for how this could look like too, similar to the original idea except it keeps the default 4:4:3 box aspect ratio:

```
v_intervals = np.vstack((self.xaxis.get_view_interval(),
                         self.yaxis.get_view_interval(),
                         self.zaxis.get_view_interval()))
mean = np.mean(v_intervals, axis=1)
deltas = np.max(np.ptp(v_intervals, axis=1)) * self._box_aspect/min(self._box_aspect)

self.set_xlim3d(mean[0] - deltas[0] / 2., mean[0] + deltas[0] / 2.)
self.set_ylim3d(mean[1] - deltas[1] / 2., mean[1] + deltas[1] / 2.)
self.set_zlim3d(mean[2] - deltas[2] / 2., mean[2] + deltas[2] / 2.)
```

image

If I had to choose one I think I would choose the last, but would welcome input on this. I don't believe matlab has prior art for 3D plots here that we can copy.

**Comment by scottshambaugh** [comment]
So to condense it, here are the three options I see:

1. Set the box aspect to (1, 1, 1), change axes limits to show equal data aspects
2. Set the box aspect to be tight around the data, change axes limits to show equal data aspects
3. Keep the box aspect to the present value, change axes limits to show equal data aspects

I would vote 3 since it encompasses 1, and perhaps 2 could be another keyword?

**Comment by scottshambaugh** [comment]
For the PR cleanliness check, I'm not sure how to squash the commits while making sure @patquem keeps credit for their initial commit.

**Comment by tfpf** [comment]
> I would vote 3 since it encompasses 1, and perhaps 2 could be another keyword? Will make that change unless there are any objections.

This is a nice idea. Could use the `adjustable` keyword argument—how about:
```
if adjustable == 'datalim': change the limits, keeping the box aspect the same
if adjustable == 'box': change the box aspect, keeping the limits the same
```
which would make it more closely resemble the behaviour of 2D axes.

**Review by tfpf — APPROVED** [review]
Looks good. Changing those gallery examples was also a nice touch!

**Comment by oscargus** [comment]
Nice discussion, which probably shows why it wasn't implemented earlier: there are lots of definitions of "equal" here...

One thing that struck me, that of course doesn't have to be part of this, is that maybe one would also like to add something like "equalyxy", "equalxz" and "equalyz" to set two out of three axes equal?

**Comment by scottshambaugh** [comment]
@oscargus good idea, wasn't too hard to add!

**Review by oscargus — APPROVED** [review]

**Comment by scottshambaugh** [comment]
Update to not shrink the plot so much:
image

**Comment by scottshambaugh** [comment]
Update to better fill the available space, changing the box to side lengths (1, 1, 5) to exaggerate what's happening:
image

**Comment by scottshambaugh** [comment]
Huh, gallery updates must have gotten lost in a rebase, added back in

**Comment by scottshambaugh** [comment]
As discussed in the dev call yesterday, this is now a blocker for https://github.com/matplotlib/matplotlib/pull/23449.

**Comment by tacaswell** [comment]
🎉 I'm very excited this is going in!

**Pull request merged**